### PR TITLE
Week 2 assessment Elmer Figueroa

### PR DESCRIPTION
**Pull request by drfudd2001**
# Submitting Week 2 Assessment

### ✅ Completed Files

Select the following as confirmation that the files in this repository are complete.

- ✅ Coding Challenges
- ✅ Coding Concepts
- ✅ Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:

I might have gone a little overboard with my answers, but I took last assessment's comments to heart.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -11,49 +11,241 @@
...
+
+
+
+
 // --------------------1) Create a function that takes a object as an argument and decides if the number inside it is evenly divisible by three or not.
 
+// Pseudocode:
+    // 1. Extract the number value from the object.
+    // 2. Check if the number is divisible by 3 using the modulo < % > operator.
+    // 3. If the result of the modulo operation is 0, return a string indicating that the number is divisible by three. Otherwise, return a string indicating that the number is not divisible by three.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:30]: great approach

File: code-challenges.test.js [hunk 2]
```diff
@@ -11,49 +11,241 @@
...
+
+
 // a) Create a test with expect statements for each of the variables provided.
 
+// Pseudocode:
+    // 1. Describe the behavior of the function using the describe function and provide a name for the test suite.
+    // 2. Use the it function to define individual test cases within the test suite.
+    // 3. For each test case, provide a description of the expected behavior of the function and the input that will be passed to the function.
+    // 4. Inside each test case, call the function with the input and use the expect function to make assertions about the output.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:39]: nice you provided pseudocode for your test

File: code-challenges.test.js [hunk 3]
```diff
@@ -11,49 +11,241 @@
...
+})
+
+    //Result:
+        // FAIL  ./code-challenges.test.js
+        // divisibleByThree
+        //   ✕ returns '15 is divisible by three' when given { number: 15 }
+        //   ✕ returns '0 is divisible by three' when given { number: 0 }
+        //   ✕ returns '-7 is not divisible by three' when given { number: -7 }
+
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:78]: solid failure

File: code-challenges.test.js [hunk 4]
```diff
@@ -11,49 +11,241 @@
...
+    // 5. Add the capitalized word to the empty array created in step 2 using the push method.
+    // 6. After the loop is finished, return the array with capitalized words.
+
 // a) Create a test with expect statements for each of the variables provided.
 
+// Pseudocode:
+    // 1. Describe the behavior of the function using the describe function and provide a name for the test suite.
+    // 2. Use the it function to define individual test cases within the test suite.
+    // 3. For each test case, provide a description of the expected behavior of the function and the input that will be passed to the function.
+    // 4. Inside each test case, call the function with the input and use the expect function to make assertions about the output.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:124]: very thorough pseudocode

File: code-challenges.test.js [hunk 5]
```diff
@@ -11,49 +11,241 @@
...
+    expect(result).toEqual(["Temperature", "Database", "Chopsticks", "Mango"])
+    })
+})
+
+    // Result:
+        // FAIL  ./code-challenges.test.js
+        // capitalizeWords
+        //  ✕ returns an array with capitalized words when given ['streetlamp', 'potato', 'teeth', 'conclusion', 'nephew']
+        //  ✕ returns an array with capitalized words when given ['temperature', 'database', 'chopsticks', 'mango']
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:152]: Thanks for inserting the test suite results

File: code-challenges.test.js [hunk 6]
```diff
@@ -11,49 +11,241 @@
...
+function capitalizeWords(words) {
+    return words.map(word => word.charAt(0).toUpperCase() + word.slice(1))
+    }
+
+    // Result:
+        // PASS  ./code-challenges.test.js
+        // capitalizeWords
+        //  ✓ returns an array with capitalized words when given ['streetlamp', 'potato', 'teeth', 'conclusion', 'nephew'] (1 ms)
+        //  ✓ returns an array with capitalized words when given ['temperature', 'database', 'chopsticks', 'mango']
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:171]: Success!!!

File: code-challenges.test.js [hunk 7]
```diff
@@ -11,49 +11,241 @@
...
+// Pseudocode:
+    // 1. Define a function named firstVowelIndex that takes a string as its argument.
+    // 2. Loop through the string using a for loop.
+    // 3. For each character in the string, check if it is a vowel.
+    // 4. If the character is a vowel, return its index.
+    // 5. If no vowels are found in the string, return -1 (Takes into account any edge cases).
+
+function firstVowelIndex(str) {
+    for (let i = 0; i < str.length; i++) {
+    if (/[aeiou]/i.test(str[i])) {
```
Comment (SunkissedQueen) [inline code-challenges.test.js:238]: interesting approach

File: code-challenges.test.js [hunk 8]
```diff
@@ -11,49 +11,241 @@
...
-const numbersArray2 = [24, 27, 30, 33, 36]
+// const numbersArray2 = [24, 27, 30, 33, 36]
 // Expected output: [72, 81, 90, 99, 108]
-
 // b) Create the function that makes the test pass.
 
+
+
+
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:24]: Looks like you have a great handle on the coding challenges. Here are some refactor suggestions.

1. Suggest trying a refactor on your test that will allow you to use less lines. 
2. Check for a method that will allow you to use less lines on the function for question 3. For questions 1, if you feel up to the challenge look into using a ternary operator.

File: coding-concepts.js [hunk 9]
```diff
@@ -10,35 +10,54 @@
 const cohort = "Bravo 2023"
 // console.log(cohort.split(""))
 
-// a) Your answer:
+// a) Your answer: [ 'B', 'r', 'a', 'v', 'o', ' ', '2', '0', '2', '3' ]
 // b) Verify and explain:
 
+    // This will log an array of each character in the string cohort. The split() method is being called on the string with an empty string as its argument, which splits the string at every character, resulting in an array where each element is a character in the original string. In this case, the resulting array will look like ["B", "r", "a", "v", "o", " ", "2", "0", "2", "3"], where each element is a single character from the original string cohort.
```
Comment (SunkissedQueen) [inline coding-concepts.js:16]: nice

File: coding-concepts.js [hunk 10]
```diff
@@ -10,35 +10,54 @@
...
 const greeter = (name) => {
   `Hello, ${name}!`
 }
 // console.log(greeter("LEARN Student"))
 
-// a) Your answer:
-// b) Verify and explain:
+// a) Your answer: undefined (it's mising the return statement)
+// b) Verify and explain: 
+
```
Comment (SunkissedQueen) [inline coding-concepts.js:27]: love that you provided solutions

File: coding-concepts.js [hunk 11]
```diff
@@ -10,35 +10,54 @@
...
 // --------------------3) What will this log?
 
 const multipliedByTwo = [4, 5, 6, 7, 8].map((number) => number * 2)
 // console.log(multipliedByTwo)
 
-// a) Your answer:
+// a) Your answer: [ 8, 10, 12, 14, 16 ]
 // b) Verify and explain:
 
+    // The function const multipliedByTwo = [4, 5, 6, 7, 8].map((number) => number * 2):
```
Comment (SunkissedQueen) [inline coding-concepts.js:47]: good job

File: coding-concepts.js [hunk 12]
```diff
@@ -49,9 +68,12 @@ const myCodingSkills = {
 }
 // console.log(myCodingSkills.languages[0])
 
-// a) Your answer:
+// a) Your answer: JavaScript
```
Comment (SunkissedQueen) [inline coding-concepts.js:71]: absolutely

File: coding-concepts.js [hunk 13]
```diff
@@ -64,5 +86,8 @@ class Learn {
 const learnStudent = new Learn("George")
 // console.log(learnStudent)
 
-// a) Your answer:
-// b) Verify and explain:
\ No newline at end of file
+// a) Your answer: Learn { student: 'George', cohort: 'Bravo', year: 2023 }
+// b) Verify and explain:
+
```
Comment (SunkissedQueen) [inline coding-concepts.js:91]: got the concept

File: interview-questions.md [hunk 14]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+-------
 
-Your answer:
+### 1. What is the difference between a parameter and an argument?
+
+Your answer: 
+
+A parameter is a variable declared in a function's definition, while an argument is the actual value passed to the function when it is called.
 
 Researched answer:
```
Comment (SunkissedQueen) [inline interview-questions.md:18]: Superb way to show that you can research, comprehend information, and share that information.

File: interview-questions.md [hunk 15]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+
+In summary, the .map() method is a powerful function in JavaScript that allows developers to easily transform each element in an array. It requires a callback function as its first parameter, which is called for each element in the array and must return a new value. The second parameter, this, is optional and is used to set the context for the callback function.
+
+Source:
+
+"JavaScript Map – How to Use the JS .map() Function (Array Method)" byy Nathan Sebhastian - https://www.freecodecamp.org/news/javascript-map-how-to-use-the-js-map-function-array-method/
```
Comment (SunkissedQueen) [inline interview-questions.md:51]: Wonderful structure. Great content for an interview

File: interview-questions.md [hunk 16]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+
+On the other hand, filter() is used when you want to create a new array from the original array, containing only the elements that pass a certain condition. It creates a new array with only the elements that pass the test function. The original array is not modified by the filter() method. It is also a pure function that does not mutate the original array. The filter() method requires a callback function as an argument, which takes the current element of the array as an argument and returns a boolean value indicating whether the element should be included in the new array.
+
+Both map() and filter() methods can be chained together with other array methods, allowing for more complex operations to be performed on arrays in a concise and readable manner. While the main difference between map() and filter() is that the former transforms each element into a new value and the latter selects only certain elements, both methods are powerful and versatile and can be used to solve a wide range of problems when working with arrays.
+
+Source: 
+
+SitePoint: "JavaScript Map vs. Filter: Simplify your JavaScript Code" - https://www.sitepoint.com/map-vs-filter-javascript/
```
Comment (SunkissedQueen) [inline interview-questions.md:73]: I suggest you start a blog with your answers. It is quite easy to share on linkedIn. I also recommend sharing with medium.com. These answers could be very helpful to new developers as well a way for you to get exposure in the web development community.

File: interview-questions.md [hunk 17]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+        let myArray = [1, 2, 3, 4, 5]
+        for (let i = 0; i < myArray.length; i++) {
+            console.log(myArray[i])
+        }
+
+In this example, the for loop starts at index 0 and iterates over the array until the index is less than the length of the array. For each iteration, the loop prints the value of the current element to the console.
+
+Source: 
+
+MDN Web Docs: Iteration - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Loops_and_iteration
```
Comment (SunkissedQueen) [inline interview-questions.md:96]: Bravo for adding an example.

File: interview-questions.md [hunk 18]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+
+        myObject.myMethod(); // call the method on the object
+
+In this example, myMethod is a method that is attached to the myObject object. The method can be called using dot notation on the object.
+
+Source: 
+
+MDN Web Docs: Functions - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Functions
+
+MDN Web Docs: Working with objects - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Working_with_Objects
```
Comment (SunkissedQueen) [inline interview-questions.md:139]: splendiferous

File: interview-questions.md [hunk 19]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+For example, the following code will log undefined because the variable is declared after it is used:
+
+        console.log(myVar); // undefined
+        var myVar = "Hello, world!"
+
+In this example, the myVar variable is declared after it is used, but it is hoisted to the top of the scope so the code still works. However, the variable is undefined until it is assigned a value.
+
+Source:
+
+MDN Web Docs: Hoisting - https://developer.mozilla.org/en-US/docs/Glossary/Hoisting
```
Comment (SunkissedQueen) [inline interview-questions.md:175]: very good explanation

File: interview-questions.md [hunk 20]
```diff
@@ -6,52 +6,416 @@ First, without external resources. Challenge yourself to answer from memory as i
...
+
+------
+
+
 ## Looking Ahead: Terms for Next Week
```
Comment (SunkissedQueen) [inline interview-questions.md:181]: Celebrating the opportunity you are giving yourself with the efforts you placed into this assessment.

**Review by SunkissedQueen — APPROVED** [review]
Such quality work! Elmer, it was a pleasure seeing the passion you poured into this assessment. You are certainly showing competent developing skills and wonderful soft skills. If you maintain this level, you will certainly not have any problems being selected for internship nor being noticed in the web development community. Suggestions for next assessment:
1. Refactor. 
Once you have accomplished the mission for coding challenges, please go back and do a refactor to remove any redundant code and/or use a different method that will reduce the amount of lines you use in your functions.
2. Add more content to interview questions.
You certainly understand vocabulary and concepts. I see that you used examples on some interview questions. That is a great way to show that can use the tool. Additional content to allow the interviewer to engage with you is to share personal experiences concerning the topic. For instance, you can share experiences while pair programming or while building your own projects.
Looking forward to your next assessment!!!
